Implement Two Factor Authentication

tsteur opened the pull request:
fixes #13325 in a new plugin TwoFactorAuth

Also provides a new class `PasswordVerifier` which requires a user to first confirm their password in a new screen before continuing with a specific action (not available in a dialog currently).

tsteur committed:
started work on two factor authentication

tsteur committed:
more work on 2fa

tsteur committed:
move code for two factor auth into a plugin

tsteur committed:
remove some no longer needed code

tsteur committed:
backup codes

tsteur committed:
some more work on 2fa

tsteur committed:
making most of the 2fa flow work

tsteur committed:
make 2fa work for logme, for the API, etc. show in users manager, do not return secret in user api etc

tsteur committed:
improve ui, increase security, etc

tsteur committed:
migrate google authenticator plugin

tsteur committed:
make process work when 2fa required but not set up yet, added command to reset 2fa

tsteur committed:
let super users reset 2fa

tsteur committed:
added translations

tsteur committed:
lots of tests, improved code, renamings, ...

tsteur committed:
started writing some ui tests, not working yet

tsteur committed:
security fixes, fixes, added more tests, ...

tsteur committed:
fix some tests

tsteur committed:
fix some more tests

tsteur committed:
fix more tests

tsteur committed:
fix more tests

tsteur committed:
more tests, fixes, and improvements

tsteur committed:
trying to fix more tests

tsteur committed:
fix more tests

tsteur committed:
fix more tests

tsteur committed:
Merge branch '3.x-dev' into 13325

tsteur committed:
fix more tests

tsteur committed:
consistent user info behaviour

tsteur committed:
fix more tests

tsteur committed:
fix more ui tests

tsteur committed:
added links

tsteur commented:
Todo: We need to update either this PR or https://github.com/matomo-org/matomo/pull/13472 depending on which one gets merged first to apply brute force / lock down changes to #13670

tsteur committed:
add learn more link

tsteur committed:
better learn more link

tsteur committed:
ensure host is trusted

diosmosis commented on plugins/Login/Controller.php:
```diff
@@ -148,6 +159,51 @@ protected function configureView($view)
+            } else {
+                $messageNoAccess = Piwik::translate('Login_WrongPasswordEntered');
+            }
+        }
```
If this method is for verifying the two factor auth code, I think it might be confusing to use the word password for both the user's password and the two factor code. Some services use "two-factor passcode", maybe we could use that?

diosmosis commented on plugins/TwoFactorAuth/Commands/Disable2FAForUser.php:
```diff
@@ -0,0 +1,36 @@
+
+        StaticContainer::get(TwoFactorAuthentication::class)->disable2FAforUser($login);
+        $message = sprintf('<info>Disabled two-factor authentication for user: %s</info>', $login);
+        $output->writeln($message);
```
Don't think it's super important, but should this go through `Request::processRequest()`?

tsteur replied:
will change it. 👍

diosmosis commented on plugins/TwoFactorAuth/Validator.php:
```diff
@@ -0,0 +1,84 @@
+    public function checkCanUseTwoFa()
+    {
+        Piwik::checkUserIsNotAnonymous();
+    }
```
Should this also have the `SettingsPiwik::isPiwikInstalled()` check?

tsteur replied:
yep, makes sense 👍  will update.

diosmosis commented on plugins/TwoFactorAuth/templates/userSettings.twig:
```diff
@@ -0,0 +1,41 @@
+            <a href="{{ linkTo({'module': 'TwoFactorAuth', 'action': 'disableTwoFactorAuth', 'disableNonce': disableNonce}) }}" style="display:none;" id="disable2fa">disable2fa</a>
+            <input type="button"
+                   class="btn btn-link disable2FaLink"
+                   onclick="piwikHelper.modalConfirm('#confirmDisable2FA', {yes: function () { broadcast.propagateNewPage('module=TwoFactorAuth&action=disableTwoFactorAuth&disableNonce={{ disableNonce|e('url') }}'); }})"
```
Should code this be in a JS file instead of the onclick attribute?

tsteur replied:
I'll move the code within `onclick` into a JS file but will likely leave the onclick itself re simplicity...

diosmosis replied:
👍

diosmosis commented on plugins/UsersManager/lang/en.json:
```diff
@@ -1,5 +1,10 @@
+        "UsesTwoFactorAuthentication": "Uses two-factor authentication",
+        "TwoFactorAuthentication": "Two-factor authentication",
+        "ResetTwoFactorAuthentication": "Reset two-factor authentication",
+        "ResetTwoFactorAuthenticationInfo": "If the user cannot log in anymore because of lost recovery codes and the authentication device you can reset two-factor authentication so the user can log in again.",
```
Translation tweak: `"If the user can no longer log in due to lost recovery codes or a lost authentication device, you can reset two-factor authentication for the user so they can log in again."`

tsteur replied:
👍  will update

diosmosis commented:
From testing locally:
* in the setupTwoFactorAuth page, if I copy text manually, it doesn't enable the 'next' button, which confused me a bit.
* not sure how important it is, but the 2 factor auth code input seems to save values:

![image](https://user-images.githubusercontent.com/125140/49127510-6b9b3380-f27c-11e8-886d-61bf166434cc.png)

seems a bit odd to see past values here.
* the mailto for the "Ask super user to reset your authentication code " link isn't set properly for non-superusers. likely due to the change in UsersManager::enrichUser (the email is removed if the user isn't the current user)

diosmosis commented:
Looked through the code, seems ok but there's a lot to review, so could have missed something. @sgiehl might want to review as well

tsteur committed:
apply review feedback

tsteur commented:
@diosmosis fyi applying all the changes you suggested.

> the mailto for the "Ask super user to reset your authentication code " link isn't set properly for non-superusers. likely due to the change in UsersManager::enrichUser (the email is removed if the user isn't the current user)

this is a bug and regressed... will also look into how to fix it...

> in the setupTwoFactorAuth page, if I copy text manually, it doesn't enable the 'next' button, which confused me a bit.

not sure if we can detect whether all the codes have been copied manually? sounds bit tricky maybe but will check if possible

tsteur committed:
Merge branch '3.x-dev' into 13325

tsteur committed:
fix emails of super users were no longer visible

tsteur committed:
fix qr code not working, fix ui tests, add activity log support

tsteur commented:
> Could just allow users to click next w/o copying/etc. Would only be important if lots of people tried to copy manually.

platforms like github force you to press one of the buttons basically to increase likelihood that they have created a backup of the recovery codes. I just checked and it does show a notification

> Please backup your recovery codes using one of the above methods before continuing the two-factor authentication setup.

So I reckon it should be good for now... also fixed couple screenshot tests but a few more need to be updated after next run. Also added support for activity log...

@sgiehl let me know whether you want to have a look or whether it can be merged.

tsteur committed:
fix more screens

tsteur committed:
Merge branch '3.x-dev' into 13325

sgiehl commented:
I will have a look now...

sgiehl commented on core/Db/Schema/Mysql.php:
```diff
@@ -52,6 +53,14 @@ public function getTablesCreateSql()
+            'twofactor_recovery_code'    => "CREATE TABLE {$prefixTables}twofactor_recovery_code (
+                          idrecoverycode BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
+                          login VARCHAR(100) NOT NULL,
+                          recovery_code VARCHAR(40) NOT NULL,
```
That one differs to the update script, it only creates a `VARCHAR(20)`

tsteur replied:
👍  will fix

sgiehl reviewed (changes requested):
Some Feedback:
- Left some comments

- When 2FA is already setup the UI imho does not make it clear enough that `Setup a different device` will also remove the currently configured 2FA and also the recovery codes. Github for example displays this warning when changing the 2FA device:
![image](https://user-images.githubusercontent.com/1579355/49344277-20669380-f675-11e8-89de-a7ffc23ba9ec.png)

- Migration from GoogleAuthenticator plugin seems to work fine. Shall I release a new version that marks the plugin not compatible with Matomo 3.8?

tsteur commented on plugins/UsersManager/lang/en.json:
```diff
@@ -1,5 +1,10 @@
 {
     "UsersManager": {
+        "2FA": "2FA",
```
I left it there separately in case TwoFactorAuth plugin is disabled. I don't think we load the translation then? We're not checking in the UI whether 2FA plugin is disabled in UsersManager as it's edge case and bit tricky. This way it still shows correct translation there.

tsteur committed:
apply more review feedback

tsteur commented:
@sgiehl now showing a warning re changing the device. Also fixed the update schema 

Yes sounds good to mark Google authenticator as invalid with 3.8 👍

tsteur committed:
Merge branch '13325' of github.com:matomo-org/piwik into 13325

tsteur committed:
Merge branch '3.x-dev' into 13325

tsteur committed:
fix ui tests

tsteur committed:
Merge branch '13325' of github.com:matomo-org/piwik into 13325
